Uniform `base_ring_type` method

fingolfin opened the pull request:
This is mostly about making `base_ring_type` more convenient to use. 

Found this this while cleaning up old branches in my local clone, apparently I made it last November, but never submitted... Huh. Not sure why, perhaps because I started to question whether it should really apply to elements, too... *shrug* you tell me what you think, I guess.

lgoettgens commented:
Nice addition! However, there are some types that do not have a `base_ring_type` method yet and cause the conformance test to fail. Furthermore, if we consider this as a required part of the ring interface (as we test it in the conformance test), the ring interface docs should mention it as well.

fingolfin commented:
I've changed this now and removed the conformance tests. This way, we can hopefully merge this right away. I'll make a separate PR with the conformance tests and documentation adjustments.
